app/presenters/tree_builder_vms_and_templates.rb

```diff
@@ -51,12 +51,14 @@ def reparent_hidden_folders(tree)
   end
 
   def prune_rbac(tree)
-    vms = extract_vms(tree).uniq
-
-    allowed_vm_ids = Set.new(Rbac.filtered(vms).map(&:id)) # results_format == :ids
+    # results_format == :ids
+    allowed_vm_ids = if User.current_user.settings[:display_vms]
+                       Set.new(Rbac.filtered(extract_vms(tree).uniq).map(&:id))
+                     else
+                       Set.new([])
+                     end
 
     prune_filtered_vms(tree, allowed_vm_ids)
-    prune_empty_folders(tree)
```

Comment:
I have solved this problem. I do not think you need to change anything in `tree_builder_vms_and_templates` - my PR should handle those